### PR TITLE
tmpfiles: make --purge more restrictive, and various other tweaks

### PR DESCRIPTION
**Pull request by poettering**
My fix for #33349, alternative for #33353 that tries to properly address things.

This attacks the problem on three fronts:

1. We refuse to run --purge without specification of at least one tmpfiles.d/ drop-in filename
2. We document in the man page much clearer what this does with `<emphasis>` and `Warning!`. 
3. We document in the --help text better what this actually does, and how it differs from --remove.

I am pretty sure with these changes it's quite hard to still fuck up your system with this, i.e. any chance to learn about --purge should already make clear to you that this is not what you want, and even if you ignore that, you have to go out your way to specify the wrong tmpfiles.d/ snippet on the command line.

Plus a multitude of other fixes, some quite important, such as one that fixes a case where --dry-run wasn't dry at all.

Fixes: #33349
Closes: https://github.com/systemd/systemd/pull/33353

### CODE REVIEW

File: src/tmpfiles/tmpfiles.c
Comment (YHNdnzj) [inline src/tmpfiles/tmpfiles.c]: I think it's better to move this up, closer to the arg_operation == 0 check?

**Review by YHNdnzj — APPROVED** [review]

**Review by bluca — APPROVED** [review]

**Comment by bluca** [comment]
Thanks, this looks right the right fix to me

**Comment by bluca** [comment]
I'll prep a stable update after this is merged

**Comment by YHNdnzj** [comment]
> I'll prep a stable update after this is merged

Please also wait for #33258 if possible. I think the accounting really should be fixed

**Comment by purpleidea** [comment]
Smalll typo: s/combinatin/combination/

**Comment by bluca** [comment]
Pushed the trivial fixes to move the check and fix the typo. Once the fast CI is done I'll merge.

**Comment by bluca** [comment]
> > I'll prep a stable update after this is merged
> 
> Please also wait for #33258 if possible. I think the accounting really should be fixed

That needs Lennart's review I think - we do stable releases regularly, it can go in the next one if it's not done in time

File: src/tmpfiles/tmpfiles.c
Comment (keszybz) [inline src/tmpfiles/tmpfiles.c:4150]: This should say "Recursively delete", unless the mechanism is changed later.

EDIT: It's not. I was reading patches in order and wasn't sure whether there's some change in behaviour later.

**Review by keszybz — COMMENTED** [review]
This was rushed through. There were points raised in the discussion in the original issue and the other PR that are not addressed. I'm a bit disappointed that you didn't even wait for my review, since this was done partially in response to my pull request.

The biggest problem I have with this that it doesn't answer the question that `--purge` most likely shouldn't recursively remove files that it doesn't specify or "own" in any way.

**Comment by bluca** [comment]
> This was rushed through. There were points raised in the discussion in the original issue and the other PR that are not addressed. I'm a bit disappointed that you didn't even wait for my review, since this was done partially in response to my pull request.

Sorry - but it was open since this morning, and approved by multiple people. I mean, either this is superurgentzomgneedastablereleaselikeyesterday, or it isn't. Seemed like to me a working day was enough, and I was already prepping the stable backport, so wanted to get this in as well.

> The biggest problem I have with this that it doesn't answer the question that `--purge` most likely shouldn't recursively remove files that it doesn't specify or "own" in any way.

It should. If I specify `d /var/mypkg/` and then `mypkg` writes a bunch of stuff underneath `/var/pkg`, it very much should remove it all when purging `mypkg`, as that's the purpose. Having to specify everything individually makes it pretty much useless.

**Comment by keszybz** [comment]
Well, in the other PR, I asked for you input because I assumed that you'll have opinions. I would expect the same courtesy to be applied here. In fact, I would expect that we'd wait for a review from the people involved in the discussion in general. We can skip that if the fix is trivial and it's obvious there's nothing left to discuss, but that's hardly the case here.

---

> If I specify `d /var/mypkg/` and then `mypkg` writes a bunch of stuff underneath `/var/pkg`, it very much should remove it all when purging `mypkg`

I'm not familiar with the debhelper tooling, so I don't know if this is opt-in or automatic.

If I say `dpkg -P systemd`, that shall remove `/home`?

**Comment by bluca** [comment]
> Well, in the other PR, I asked for you input because I assumed that you'll have opinions. I would expect the same courtesy to be applied here. In fact, I would expect that we'd wait for a review from the people involved in the discussion in general. We can skip that if the fix is trivial and it's obvious there's nothing left to discuss, but that's hardly the case here.

Again, sorry about that - I thought we wanted to get out 256.1 with the belt and braces asap

> > If I specify `d /var/mypkg/` and then `mypkg` writes a bunch of stuff underneath `/var/pkg`, it very much should remove it all when purging `mypkg`
> 
> I'm not familiar with the debhelper tooling, so I don't know if this is opt-in or automatic.
> 
> If I say `dpkg -P systemd`, that shall remove `/home`?

`apt remove --purge systemd` (which would ask you to remove most of your system, so seems hard to do accidentally) and no, the automation is not plugged in the systemd binary package, so it wouldn't

**Comment by keszybz** [comment]
> Again, sorry about that - I thought we wanted to get out 256.1 with the belt and braces asap

OK.

> `apt remove --purge systemd` (which would ask you to remove most of your system, so seems hard to do accidentally) and no, the automation is not plugged in the systemd binary package, so it wouldn't. It's automatic _if_ the package maintainer scripts and d/rules are set up in a certain way, that should be the default but it's not universal, and there's an opt out

OK, but how is it set up today. If `systemd.deb` is rebuilt with `systemd-tmpfiles` that supports purge, will `apt remove --purge systemd` remove `/home`?

**Comment by bluca** [comment]
> > `apt remove --purge systemd` (which would ask you to remove most of your system, so seems hard to do accidentally) and no, the automation is not plugged in the systemd binary package, so it wouldn't. It's automatic _if_ the package maintainer scripts and d/rules are set up in a certain way, that should be the default but it's not universal, and there's an opt out
> 
> OK, but how is it set up today? If `systemd.deb` is rebuilt with `systemd-tmpfiles` that supports purge, will `apt remove --purge systemd` remove `/home`?

Again no, as mentioned the automation is not enabled for that package. You can download the package from unstable/testing and check the `postrm` script, it's not there

**Comment by keszybz** [comment]
The helper is in https://salsa.debian.org/debian/debhelper/-/blob/main/dh_installtmpfiles?ref_type=heads…
IIUC, `dh_installtmpfiles` is the standard way to handle a tmpfile.d snippet in a package. And unless the maintainer opts-out, `systemd-tmpfiles purge` will be hooked up for `apt remove --purge`. Sorry, but this doesn't seem like a good idea _at all_.

In some way, I think this is a Debian problem, because other systems don't have the concept of "purge". Rpm certainly doesn't. This means that users would have to invoke `systemd-tmpfiles --purge` directly. This is still possible, and I'm not happy about this possibility, but it's less likely.

**Comment by bluca** [comment]
> The helper is in [https://salsa.debian.org/debian/debhelper/-/blob/main/dh_installtmpfiles?ref_type=heads…](https://salsa.debian.org/debian/debhelper/-/blob/main/dh_installtmpfiles?ref_type=heads%E2%80%A6) IIUC, `dh_installtmpfiles` is the standard way to handle a tmpfile.d snippet in a package. And unless the maintainer opts-out, `systemd-tmpfiles purge` will be hooked up for `apt remove --purge`. Sorry, but this doesn't seem like a good idea _at all_.

Yes, that is all correct, as mentioned it's opt-out - in the new compat mode of course as it's an incompatible change, and it is documented. So one has to first switch from debhelper-compat 13 to 14 to get the new behaviour. Purging everything that is left behind is the reason this exists in the first place.

> In some way, I think this is a Debian problem, because other systems don't have the concept of "purge". Rpm certainly doesn't. This means that users would have to invoke `systemd-tmpfiles --purge` directly. This is still possible, and I'm not happy about this possibility, but it's less likely.

Yeah but just invoking it does nothing now, which is the point - the belt and braces. You have to _explicitly_ call `systemd-tmpfiles --purge /usr/lib/tmpfiles.d/home.conf`, which with the updated documentation doesn't seem any different to me than doing `rm -rf /home`. I mean, it's even worse, because running `rm -rf` is a common operation if someone uses the terminal, while let's be real, absolutely nobody would habitually call sd-tmpfiles manually in the real world. I mean, as far as I'm concerned we could even move it out of bin and place it with the other private binaries that are run by services in /usr/lib/systemd.

**Comment by keszybz** [comment]
> Yes, that is all correct, as mentioned it's opt-out - in the new compat mode of course as it's an incompatible change, and it is documented. So one has to first switch from debhelper-compat 13 to 14 to get the new behaviour. Purging everything that is left behind is the reason this exists in the first place.

That still seems to be a very bad idea. I'm pretty sure not all packagers read all debhelper-compat changelogs attentively, and even if they did, somebody might add a new tmpfiles line at some later point, either upstream or downstream.

> You have to _explicitly_ call `systemd-tmpfiles --purge /usr/lib/tmpfiles.d/home.conf`

`systemd-tmpfiles --purge home.conf` is enough. But I'll concede that invoking it manually like that seems unlikely to be done accidentally. But invocations with `--purge` are wired-up automatically by `dh_installtmpfiles`, i.e. they will be called from `apt purge`. *That* seems very risky.

I really don't have a problem with `tmpfiles` removing a bunch of files or temporary directories. But the recursive operation on stateful directories I cannot get behind.

**Comment by bluca** [comment]
> > Yes, that is all correct, as mentioned it's opt-out - in the new compat mode of course as it's an incompatible change, and it is documented. So one has to first switch from debhelper-compat 13 to 14 to get the new behaviour. Purging everything that is left behind is the reason this exists in the first place.
> 
> That still seems to be a very bad idea. I'm pretty sure not all packagers read all debhelper-compat changelogs attentively, and even if they did, somebody might add a new tmpfiles line at some later point, either upstream or downstream.

I mean, a package also might add a unit that does dd if=/dev/zero of=/dev/sda on boot - being a packager means auditing what you package, and means reading the documentation of the tools that you use, this is not really anything new, nor it is a convincing reason for not integrating it. If the possibility of some unknown hypothetical person opting in to some functionality without reading what it does would be an issue, then nothing could be done, ever.

> > You have to _explicitly_ call `systemd-tmpfiles --purge /usr/lib/tmpfiles.d/home.conf`
> 
> `systemd-tmpfiles --purge home.conf` is enough. But I'll concede that invoking it manually like that seems unlikely to be done accidentally. But invocations with `--purge` are wired-up automatically by `dh_installtmpfiles`, i.e. they will be called from `apt purge`. _That_ seems very risky.
> 
> I really don't have a problem with `tmpfiles` removing a bunch of files or temporary directories. But the recursive operation on stateful directories I cannot get behind.

But I'm not asking you to add this to RPM - as you already said, RPM doesn't really have the concept of 'purging', so it doesn't really apply anyway. So what's the problem? You are not convinced that it's good to have, so you won't use it. I am, so I will. Everybody wins?

**Comment by keszybz** [comment]
We'll just have to agree to disagree. I'll just say that I think `--purge` in the upstream form is a somewhat-iffy-but-essentially-unlikely-to-be-problematic, but in the `apt purge` form is a hole waiting for a foot.

In Fedora, I pushed the following two commits:
- https://src.fedoraproject.org/rpms/systemd/c/a76669ee222fc4374f14f7e2a6e9a28c5ad87c0e, which drops `tmpfiles.d/home.conf`. When looking at this, I realized that this file can never have an effect when delivered from the rpm, because the rpm requires filesystem, and filesystem always creates both `/home` and `/srv`.
- https://src.fedoraproject.org/rpms/systemd/c/b3e1d52cb4f69520fa5d2679d68549b7c24ecbca, which requires `--destroy-data` if `--purge` is used without `--dry-run`. If people holler that this breaks their workflows, this can be reconsidered. For now I think it's better to be cautious.

**Comment by bluca** [comment]
We could also drop it from the manpage if you like, I'm ok with that - as mentioned, I think this is for scripted use anyway

**Comment by keszybz** [comment]
The man page now has an explanation and warnings, so I think it's better to keep it.

**Comment by poettering** [comment]
> * https://src.fedoraproject.org/rpms/systemd/c/a76669ee222fc4374f14f7e2a6e9a28c5ad87c0e, which drops `tmpfiles.d/home.conf`. When looking at this, I realized that this file can never have an effect when delivered from the rpm, because the rpm requires filesystem, and filesystem always creates both `/home` and `/srv`.

Oh man, this really sucks. I build mkosi images regularly that only contain a /usr/ and that augment the root fs via tmpfiles.d/. If you drop this then you are actively breaking my setup. Uncool. And why even? I don't get it. 

Please put that back, there's a reason this exists and I use it regularly. It's also what makes `systemd-nspawn --volatile=full --bind-user=…` work and things.

**Comment by poettering** [comment]
You know, I am happy to just kill the --purge thing entirely again, if this means you are breaking systemd that come up with /usr/ only. I find the purge usecase much weaker than the usecase of systems with /usr/ only.

Kinda sucks that the /usr/ merge is kinda pointless in Fedora now, if you cannot actually deploy a system anymore with just /usr/...

**Comment by poettering** [comment]
Hmm, so the rpm doesn't ship sysusers.d/ stuff either anymore? Man this is so disappointing. I wasn't aware Fedora is going backwards so badly.

**Comment by bluca** [comment]
Join the Debian dark side. We have cookies :-P

**Comment by poettering** [comment]
> This was rushed through. There were points raised in the discussion in the original issue and the other PR that are not addressed. I'm a bit disappointed that you didn't even wait for my review, since this was done partially in response to my pull request.

I actually agree. I don't think there was any need to hurry this in like this, not even giving me the chance to address @YHNdnzj review points myself.

**Comment by poettering** [comment]
btw, one way out for this would be to add yet another line type modifier to things. For example, we could say that if you combine some line type with `o` (for "ownership") it will be subject to `--purge`, but nothing else. I think that would be reasonably safe.

Example:

```
# create a dir, and make it subject to --purge
do /run/foobar 0755 root root
# create a dir, but make --purge have no effect
d /run/waldo 0755 root root
# just declare ownership
o /run/piff
```

or so.

**Comment by bluca** [comment]
> > * https://src.fedoraproject.org/rpms/systemd/c/a76669ee222fc4374f14f7e2a6e9a28c5ad87c0e, which drops `tmpfiles.d/home.conf`. When looking at this, I realized that this file can never have an effect when delivered from the rpm, because the rpm requires filesystem, and filesystem always creates both `/home` and `/srv`.
> 
> Oh man, this really sucks. I build mkosi images regularly that only contain a /usr/ and that augment the root fs via tmpfiles.d/. If you drop this then you are actively breaking my setup. Uncool. And why even? I don't get it.
> 
> Please put that back, there's a reason this exists and I use it regularly. It's also what makes `systemd-nspawn --volatile=full --bind-user=…` work and things.

We could just add it to the local mkosi config in the repo? If that's how you use it, it wouldn't make a difference whether it came from the package or from mkosi.extra, no?

**Comment by bb010g** [comment]
> btw, one way out for this would be to add yet another line type modifier to things. For example, we could say that if you combine some line type with `o` (for "ownership") it will be subject to `--purge`, but nothing else. I think that would be reasonably safe.

Reading through the current tmpfiles.d\(5\) manual page, the following jumps out at me as a case where `o` should *not* be used:
```tmpfiles.d
# Modify sysfs but don't fail if we are in a container with a read-only /proc
w- /proc/sys/vm/swappiness - - - - 10
```

`systemd-tmpfiles --purge` should generally not try to delete `/proc/sys/` files.

As a developer, I would appreciate this as a way to communicate, via upstream tmpfiles.d\(5\) files, to distribution maintainers what files and directories the project expects to (not) own. The [__FILES__ section of Unix manuals](<https://manpages.bsd.lv/mdoc.html>) is already available for this purpose, but tmpfiles.d\(5\) files are more likely to stay in sync as code updates than manuals. It would also make me more confident about using tmpfiles.d\(5\) to create files a project doesn't own now that I know `systemd-tmpfiles --purge` exists and is likely to be used by maintainers in uninstallation hooks.

As a distribution maintainer, I could also see this being useful for split packages. For example:
```tmpfiles.d
# /usr/lib/tmpfiles.d/hello.conf
o /var/spool/hello

# /usr/lib/tmpfiles.d/hello-foo.conf
d /var/spool/hello 0755 root root

# /usr/lib/tmpfiles.d/hello-bar.conf
d /var/spool/hello 0755 root root
```
The package `hello` does not spool. Its dependencies `hello-foo` and `hello-bar` both spool to the same directory, but removing `hello-foo` while `hello-foo` is still installed *must not* purge the entire spool directory. Removing `hello` means `hello-foo` and `hello-bar` must also be removed, so the entire spool directory *can* be purged when `hello` is uninstalled.

**Pull request merged**
